src/background.js

```diff
@@ -0,0 +1,142 @@
+const setLoadingState = async (active) => {
+  const activeCss = `.caal-loading {
+    display: flex;
+  }`;
+
+  const inactiveCss = `.caal-loading {
+    display: none;
+  }`;
+
+  browser.tabs.insertCSS({ code: active ? activeCss : inactiveCss });
+};
+
+const fetchJSON = async (url) => {
+  try {
+    const response = await fetch(url);
+    if (!response.ok)
+      throw new Error("Error when attempting to fetch resource.");
+    return response.json();
+  } catch (error) {
+    sendLogMessage(error);
+    return null;
+  }
+};
+
+// Portable Game Notation (PGN) is a standard plain text format for
+// recording chess games, which can be read by humans and is also
+// supported by most chess software
+const getPGN = async (playerName, gameUrl) => {
+  try {
+    const archives = (
+      await fetchJSON(
+        `https://api.chess.com/pub/player/${playerName}/games/archives`
+      )
+    ).archives;
+
+    // Archives are given from the oldest to the most recet games.
+    // Thus iterating from the last one (the most recent) is usually
+    // better, because players usually want to analyse their most recent
+    // games. So you will make less fetches
+    for (let i = archives.length - 1; i >= 0; i--) {
+      const games = (await fetchJSON(archives[i])).games;
+      for (let j = games.length - 1; j >= 0; j--) {
+        if (games[j].url === gameUrl) {
+          return games[j].pgn;
+        }
+      }
+    }
+  } catch (error) {
+    sendLogMessage(error);
+    return null;
+  }
+
+  return null;
+};
+
+const sendLogMessage = async (message) => {
+  await browser.tabs.executeScript({
+    code: `console.log("[Chess.com analyse at lichess]: ${message}");`,
+  });
+};
+
+const clickElement = async (querySelector) => {
+  await browser.tabs.executeScript({
+    code: `document.querySelector("${querySelector}").click();`,
+  });
+};
+
+const waitForElement = async (querySelector) => {
+  const elementFoundScript = `document.querySelector("${querySelector}") !== null;`;
+  const retryDelay = 100;
+
+  let timeLeft = 5000;
+  let [isElement] = await browser.tabs.executeScript({
+    code: elementFoundScript,
+  });
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise((r) => setTimeout(r, retryDelay));
+    [isElement] = await browser.tabs.executeScript({
+      code: elementFoundScript,
+    });
+    timeLeft -= retryDelay;
+  }
+
+  if (!isElement) sendLogMessage(`Cannot find \`${querySelector}\` element!`);
+
+  return isElement;
+};
+
+const waitAndClick = async (querySelector) => {
+  if (await waitForElement(querySelector)) await clickElement(querySelector);
+};
+
+let analysingState = false;
```

Comment:
Cool approach :+1: Solves the problem

Minor:

What if the user wants to execute the extension for multiple games while the previous execution is still going :smile: ? I believe it could be considered as part of https://github.com/zeraye/chess-analysis-firefox-extension/issues/3. If you will allow users to change tabs, then IMO each tab should have its own `analysingState`, so the user can analyse multiple games in parallel

Reply:
I will implement it in `v1.7`